### PR TITLE
Support collect_set RESPECT NULLS [databricks]

### LINKED ISSUE #15219: [AI-AUDIT] [Spark 4.2] collect_set RESPECT NULLS (SPARK-55533)
### Description
Spark 4.2 (SPARK-55533) adds `IGNORE NULLS` / `RESPECT NULLS` clause support to `collect_set`, mirroring SPARK-55256 for `collect_list`/`array_agg` (tracked for the plugin in #15011). `CollectSet` gains a new constructor parameter `ignoreNulls: Boolean = true`; with `RESPECT NULLS` the CPU includes a single `null` element in the result set and `CollectSet.dataType` becomes `ArrayType(child.dataType, containsNull = true)`.

The plugin's `CollectSet` meta and `GpuCollectSet` are unaware of the new field:
- `GpuOverrides.scala:3850-3851` — `convertToGpu` builds `GpuCollectSet(childExprs.head, c.mutableAggBufferOffset, c.inputAggBufferOffset)`, dropping `ignoreNulls`; `tagAggForGpu()` (3841-3848) has no `ignoreNulls` guard.
- `aggregateFunctions.scala:116-138,1976-1984` — `CudfCollectSet`, `CudfMergeSets`, and the window `RollingAggregation.collectSet` all hardcode `NullPolicy.EXCLUDE`.
- `aggregateFunctions.scala:1906,2006` — `GpuCollectBase.dataType` and `CpuToGpuCollectBufferTransition.dataType` hardcode `containsNull = false`, mismatching the new CPU dataType/serialized buffer layout when `ignoreNulls = false`.

**Effect once a spark420 shim exists**: `collect_set(x) RESPECT NULLS` would run on GPU and silently omit `null` from results (CPU includes it) — incorrect results. Default `collect_set(x)` and explicit `IGNORE NULLS` are unchanged and remain correct. No currently supported Spark version is affected (the commit is only in unreleased Spark 4.2.0).

### Recommended fix (two tracks, same as #15011)
1. **Immediate (with spark420 shim bring-up)**: in the `CollectSet` meta's `tagAggForGpu()`, fall back to CPU when `!c.ignoreNulls`:
   ```scala
   if (!c.ignoreNulls) {
     willNotWorkOnGpu("collect_set(RESPECT NULLS) is not yet supported on GPU")
   }
   ```
2. **Follow-on GPU support**: plumb `ignoreNulls` through `GpuCollectSet`, select `NullPolicy.INCLUDE` in `CudfCollectSet`/`CudfMergeSets`/window aggregation when `ignoreNulls = false` (cuDF's `NullEquality.EQUAL` already dedups nulls, matching CPU set semantics), and derive `containsNull = !ignoreNulls` in `GpuCollectBase.dataType` and `CpuToGpuCollectBufferTransition.dataType`. Fix jointly with `collect_list` (#15011) since both share `GpuCollectBase` and the buffer converters. Add IGNORE/RESPECT NULLS integration tests.

### Source
- Spark commit: `1469e0eea526bea54acd59684cdf5615af8d71da` ([SPARK-55533][SQL] Support IGNORE NULLS / RESPECT NULLS for collect_set)
- JIRA: https://issues.apache.org/jira/browse/SPARK-55533
- Applicable Spark versions: 4.2.0 (GA unreleased; master + branch-4.2 + branch-4.x; only 4.2.0 preview/RC tags contain it)
- Related: SPARK-55256 / #15011 (collect_list/array_agg RESPECT NULLS)

### PR DESCRIPTION
**Pull request by firestarman**
Fixes #15219.

### Description

- Added Spark 4.2 `collect_set RESPECT NULLS` support on GPU by plumbing `ignoreNulls` through shims, expression metadata, GPU aggregate data types, CPU buffer conversion, and cuDF null policy so GPU results preserve one null like Spark CPU.
- Added Spark 4.2 group-by, global reduction, rolling-window, and fully unbounded partitioned-window integration coverage for `collect_set IGNORE NULLS` and `collect_set RESPECT NULLS` because the new SQL clause changes null handling semantics.
- Validated the Spark 4.2 jars and focused integration tests with `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=400 -Dcuda.version=cuda13 -DskipTests validate`, `mvn -s ~/.m2/settings_art.xml -Dbuildver=330 -Dcuda.version=cuda13 -DskipTests validate`, `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -pl dist -am -Dbuildver=420 -Dcuda.version=cuda13 -DskipTests -Dmaven.javadoc.skip=true package`, `SPARK_HOME=/path/to/spark-4.2.0-bin-hadoop3 TESTS="hash_aggregate_test.py window_function_test.py" TEST="test_hash_reduction_collect_set_respect_nulls or test_window_aggs_for_rows_collect_set_respect_nulls" TEST_PARALLEL=0 ./run_pyspark_from_build.sh`, and earlier focused coverage for group-by and fully unbounded windows.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

**Comment by firestarman** [comment]
build

**Review by res-life — COMMENTED** [review]
The implementation paths for `RESPECT NULLS` need two focused regression tests before this is ready: one for global reduction and one for floating-point rolling windows.

### 1. SHOULD FIX

**File and line:** `integration_tests/src/main/python/hash_aggregate_test.py:898`

**Issue:** The new `collect_set(... RESPECT NULLS)` coverage only exercises grouped and window aggregations. It does not exercise the distinct `reductionAggregate` path changed at `aggregateFunctions.scala:126` and `aggregateFunctions.scala:129`, so a regression that restores `NullPolicy.EXCLUDE` only for global reductions would still pass the current tests.

**How to fix:** Add a Spark 4.2 test alongside `test_hash_reduction_collect_set` that runs without `GROUP BY` or `OVER` and compares `IGNORE NULLS` with `RESPECT NULLS`. Cover both reduction branches with explicitly typed inputs: an integral case with duplicate null and non-null values, and a `DOUBLE` case with duplicate nulls and duplicate NaNs. Include an all-null expression in the reduction result so the null-preserving behavior is asserted directly.

**Confidence:** CERTAIN

### 2. SHOULD FIX

**File and line:** `integration_tests/src/main/python/window_function_test.py:2458`

**Issue:** `test_window_aggs_for_rows_collect_set_respect_nulls` infers the collected column as integral, while `GpuCollectSet.windowAggregation` has a separate `FloatType | DoubleType` branch at `aggregateFunctions.scala:1998-2000`. Existing floating-point rolling tests use the default `IGNORE NULLS`, so they do not protect the new null policy in that branch.

**How to fix:** Parameterize the focused rolling-window `RESPECT NULLS` test over `INT`, `FLOAT`, and `DOUBLE`, explicitly casting the collected column before creating the temporary view. For the floating-point cases, include nulls and duplicate NaNs so the test covers both null preservation and the branch's NaN de-duplication behavior.

**Confidence:** CERTAIN

**Comment by firestarman** [comment]
Addressed in `137f1badf`.

- Added `test_hash_reduction_collect_set_respect_nulls` for global reductions, covering integral inputs, `DOUBLE` inputs with duplicate nulls and duplicate NaNs, and all-null `RESPECT NULLS` behavior.
- Parameterized the rolling-window `collect_set RESPECT NULLS` test over `INT`, `FLOAT`, and `DOUBLE`; the floating-point cases include duplicate NaNs and nulls.
- Fixed the reduction implementation for `NullPolicy.INCLUDE` by using the group-by collect-set path for single-group reductions when needed, preserving nulls for all-null and floating-point reduction cases.

Validation:
- `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=400 -Dcuda.version=cuda13 -DskipTests validate`
- `mvn -s ~/.m2/settings_art.xml -Dbuildver=330 -Dcuda.version=cuda13 -DskipTests validate`
- `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -pl dist -am -Dbuildver=420 -Dcuda.version=cuda13 -DskipTests -Dmaven.javadoc.skip=true package`
- `SPARK_HOME=/path/to/spark-4.2.0-bin-hadoop3 TESTS="hash_aggregate_test.py window_function_test.py" TEST="test_hash_reduction_collect_set_respect_nulls or test_window_aggs_for_rows_collect_set_respect_nulls" TEST_PARALLEL=0 ./run_pyspark_from_build.sh` passed: 4 passed.

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala [hunk 1]
```diff
@@ -115,28 +115,46 @@ class CudfMergeLists(override val dataType: DataType) extends CudfAggregate {
...
   }
   override val name: String = "CudfCollectSet"
+
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (col: cudf.ColumnVector) => {
+      if (nullPolicy == NullPolicy.INCLUDE && col.getRowCount > 0) {
+        // cuDF reduction collectSet currently drops nulls for some INCLUDE cases. Use the
+        // group-by implementation for single-group reductions to preserve Spark's null semantics.
+        withResource(Scalar.fromInt(0)) { keyScalar =>
+          withResource(ColumnVector.fromScalar(keyScalar, col.getRowCount.toInt)) { keys =>
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala]: Addressed in `5f995a992`.

The reduction workaround now converts `col.getRowCount` with `Math.toIntExact` and reuses the exact `rowCount` when materializing the dummy key column. This avoids silent Long-to-Int overflow for very large global reductions.

Validation:
- `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=400 -Dcuda.version=cuda13 -DskipTests validate`
- `mvn -s ~/.m2/settings_art.xml -Dbuildver=330 -Dcuda.version=cuda13 -DskipTests validate`
- `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -pl dist -am -Dbuildver=420 -Dcuda.version=cuda13 -DskipTests -Dmaven.javadoc.skip=true package`
- `SPARK_HOME=/path/to/spark-4.2.0-bin-hadoop3 TESTS="hash_aggregate_test.py" TEST="test_hash_reduction_collect_set_respect_nulls" TEST_PARALLEL=0 ./run_pyspark_from_build.sh`

**Review by res-life — APPROVED** [review]
LGTM

**Comment by firestarman** [comment]
build
